### PR TITLE
Gianpiero: implemented find neighbours and has cell, all test pass

### PR DESCRIPTION
**Pull request by GianpieroCea**

### CODE REVIEW

File: hexgame/src/board.py [hunk 1]
```diff
@@ -67,30 +67,94 @@ def place_stone(self, i:int, j:int, color : Color) -> 'Board':
...
+                )
+
+
+    def __repr__(self) -> str:
+        pass
+
 
 
 if __name__ == "__main__":
+    #TODO when happy these all work move to the test file
```
Comment (dinies) [inline hexgame/src/board.py:133]: Why not moving them there right away ? In a unit test or more ?

File: hexgame/tests/test_board.py [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
+from hexgame.src.board import Board 
+from hexgame.src.cell import Cell 
 import pytest
 
 
 class TestBoardProperties:
 
     def test_create_empty_board(self):
-        empty_board = Board.Board(0, 0)
+        empty_board = Board(0, 0)
```
Comment (dinies) [inline hexgame/tests/test_board.py:9]: I like it more this way! Nice.

File: hexgame/tests/test_board.py [hunk 3]
```diff
@@ -34,9 +35,9 @@ def test_neighbours_of_cell(self):
         assert Cell(1, 1) in from_left_edge
         assert Cell(1, 2) in from_left_edge
         assert Cell(0, 2) in from_left_edge
-        mediumBoard = Board.Board(3, 3)
+        mediumBoard = Board(3, 3)
         from_centre = mediumBoard.find_neighbours((1, 1))
-        assert len(from_left_edge) == 6
+        assert len(from_centre) == 6
```
Comment (dinies) [inline hexgame/tests/test_board.py:40]: Oh I missed that. :D

File: hexgame/tests/test_board.py [hunk 4]
```diff
@@ -1,24 +1,25 @@
-from hexgame.src import board as Board
-from hexgame.src import cell as Cell
+from hexgame.src.board import Board
```
Comment (dinies) [inline hexgame/tests/test_board.py:1]: You are introducing a space at the end of the line here.

**Review by dinies — CHANGES_REQUESTED** [review]
Everything looks good, maybe there is something missing in `find_neighbours`.

Is it too soon to enforce `Flake8` on every file? We can address that in a follow-up PR.

**Comment by GianpieroCea** [comment]
Hey . So in find neighbors I think it should be complete (I notice I left out a Todo comment but I should have removed it).

Adding more test will reveal if there are edge cases missing, it is possible.

As for flake8, I know exactly how to set it up on vscode but I didn't bother because it was annoying me, so yeah I can try to enforce it for next one, is that I have sort of juggled different Dev environment, and now i am actually just using conda :D

File: hexgame/src/board.py [hunk 5]
```diff
@@ -67,30 +67,94 @@ def place_stone(self, i:int, j:int, color : Color) -> 'Board':
...
+                    return set((Cell(0,self.dim_y-2),Cell(1,self.dim_y-1)))
+                else:
+                    #left border inside
+                    return set((Cell(0,y-1),Cell(0,y+1),Cell(1,y),Cell(1,y+1)))
+
+            elif x == self.dim_x - 1:
+                if y == 0:
+                    #bottom right
+                    return set((Cell(self.dim_x - 2,0),Cell(self.dim_x - 1,1)))
+                elif y == (self.dim_y -1):
```
[the remark this answers is not in the document]
Comment (GianpieroCea) [reply hexgame/src/board.py:111]: Shit you are right. I will fix that and tests too maybe

[the remark this answers is not in the document]
Comment (GianpieroCea) [reply hexgame/src/board.py:112]: done,there were also top and bottom missing!

File: hexgame/src/board.py (hunk 1 above)
Comment (GianpieroCea) [reply hexgame/src/board.py:189]: I agree, can we do this in another pr? i think we can have another one where we add more, test,make it look nicer etc

File: hexgame/tests/test_board.py (hunk 4 above)
Comment (GianpieroCea) [reply hexgame/tests/test_board.py:1]: yeah if this is a flake8 thing I am not noticing

**Comment by GianpieroCea** [comment]
Okay I have only completed the find neighbours method. As for flake8 and moving the test in unit test I would in a more apt PR?

File: hexgame/src/board.py [hunk 6]
```diff
@@ -103,17 +104,72 @@ def find_neighbours(self, coords: tuple[int,int]) -> set[Cell]:
...
+                    #right border inside
+                    return set(
+                        (
+                            Cell(self.dim_x - 2,self.dim_y-2),
+                            Cell(self.dim_x - 2,self.dim_y-1),
+                            Cell(self.dim_x - 1,self.dim_y-2),
+                            Cell(self.dim_x - 1,self.dim_y-2),
+                        )
+                    )
+            elif y == 0:
```
Comment (dinies) [inline hexgame/src/board.py:131]: I don't like that we are introducing all of this redundant code.
Also the function became very long, can we keep it shorter ?
I am thinking that also we are not considering when we are searching a neighbour in a very small board that will not have all the tiles around the target coord in the same way it happens from the 3 by 3 case and bigger boards.
I suggest that we check that all the things that we are returning in the set are part of the board  ( there is a function that tells us that `has_cell()` ).
For the if else conditions, we can say that there are only up to 9 different cases possible.
Therefore I would do:
if x ==0 and y==0 :   # bottom left
elif  x==0 and y==max-1 :  # top left
...
for the 8 cases in the border and then 
else :    # case in the middle

In this way we avoid nested code  as much as possible.
What do you think ?

**Review by dinies — CHANGES_REQUESTED** [review]
I am not convinced on the current implementation of `find_neighbours`

File: hexgame/src/board.py (hunk 6 above)
Comment (GianpieroCea) [reply hexgame/src/board.py:131]: Yes, very fair.

As I writing and copy pasting I was telling to myself I just wanted to close the pr and get something working so that for example we could have added test and then refactor later , but I guess there is no reason as why that cannot be done all in once now.

And ah, that makes sense then as why you made that has cell. I also had thought of making it mushc simpler by creating the nbd just by adding a neighbour cell provided it exist.

Comment (GianpieroCea) [reply hexgame/src/board.py]: Done this now, I have: rewritten the unit test to be much simple using fixtures.
Then I have reimplemented the find neighbours functions in a much easier way, test all pass.
Along the way I also aoutformatted with pep8, sorry about that we said we were going to do in another PR

**Comment by GianpieroCea** [comment]
Done this now, I have: rewritten the unit test to be much simple using fixtures.
Then I have reimplemented the find neighbours functions in a much easier way, test all pass.
Along the way I also aoutformatted with pep8, sorry about that we said we were going to do in another PR

**Review by dinies — APPROVED** [review]
Looks amazing now !!
The tests are way more concise.
Great 💪
